### DISCUSSION TITLE
Disable Composer self-update in Windows

### DISCUSSION THREAD

**Discussion by kulkarnirahul17** [composer/composer#11265] [Q&A / Support]
Hi 
My organization wants to know if disabling composer self-update command is possible on Windows machine.
They want only administrators to validate and perform updates on individual desktop machines. 
Is there a way to disable automatic updates during Windows installations or disable composer self-update entirely unless the user is an administrator?

**Comment by Seldaek** [comment]
You could make the composer.phar owned by the admin and not writable by non-admins I suppose that'd prevent self-update from working at least?

**Reply by Seldaek** [reply]
I don't see much value in preventing people from keeping their software up to date so I don't really see a need to spend time making this any easier sorry.

**Reply by stof** [reply]
For those solutions, you might expose a `composer` script that is not directly the phar but a wrapper around it, which would block the `self-update` command.

**Reply by uuf6429** [reply]
I'm not writing on behalf of composer devs.

Wouldn't such an option be a bit stupid when you allow people to overwrite the executable anyway? Sounds to me like you're trying to make it difficult for people to update (instead they would end up manually doing it anyway - and complain about the ridiculous restrictions).

What he meant, by the way, is, to write a script (bash or PHP or whatever) that calls the original composer binary with the same arguments while excluding `selfupdate` or `self-update`....basically a hack for a hack..

**Comment by BSarmady** [comment]
Let me try and explain why someone would want to disable updates.

A hacker manages to gain access to DNS server of the company, using DNS hijacking, forwards composer update url to their own machine with one or many versions higher than current and then waits for next morning when developer opens their machine running phpStorm for example, which automatically runs the composer, now composer will update itself to new one.

From this point, hacker can remove the DNS entry and revert DNS server back to original, composer will not update ever again since it is newer version and hacker has a free backdoor to that PC and probably to entire network without even needing to gain physical access to those PCs.

Now I leave rest of it to your imagination, it can download other executable, create startup scripts ... and suddenly you see source code of a software in internet before it is even published.

To stop this kind of attack, enterprises specially finance and banking, limit access to Internet for dev environment. Infra guys get frustrated checking firewall logs to see some apps persistently keep making connection to internet and there is no way to tell them stop! So they block the entire access to Internet, inflicting headaches to developer who are trying to reach Stackoverflow for example.


In a nutshell, all apps should have a way to stop them from accessing online resources if required.

**Reply by uuf6429** [reply]
While the hacking attempt is entirely plausible, I still believe there are a few wrong things about the idea in general:
1. Composer does not update automatically, it is the users' decision. If users decide to (intentionally) install a backdoored/vulnerable package, it's not composer's job to stop them, (in fact, it might help digital investigators to actually not magically block this). Same thing applies to self-updates.
2. If enterprises are blocking potentially malicious requests at the user-land application level like you described, they should really review their security practices. In your scenario, even if composer has such blocking functionality, users could either download a different version or patch up the current one (PS: phars are basically zips, and the original source is typically not even minified).
3. Additionally, if self-updating (which by itself is not commonly used) is blocked, there's still the same problem you described with installing/updating packages. If managing packages is also blocked, then composer essentially becomes useless.
4. As mentioned in previous comments, the sure-fire way is to make the composer binary non-writable. That way users and composer itself would not be able to modify it. Keep in mind that the upgrade process does not execute remote code (as far as I am aware).
   This should be something that your sysadmins should already be able to do when provisioning your dev machine.
6. > In a nutshell, all apps should have a way to stop them from accessing online resources if required.
   
   You're proposing a complex* feature for all apps (countless hours of development) that basically cripples app functionality and which has the potential to create a lot of bugs; whose functionality is very questionable and applies to a very limited amount of users. That's a very hard sell.
7.  *I've no idea how you would configure it securely without letting users to bypass it and without relying on some sort of external config management tool.

TL;DR: the hacking scenario is plausible, but the solution you suggest is, in my opinion, doing it the wrong way (involves a lot of development effort and is actually a weak precaution).

**Reply by Seldaek** [reply]
Another thing to consider is that self-update verifies the downloaded file has a valid signature using our public key for verification, and that gets stored on initial install of composer so if dns is compromised self-update should be perfectly safe and fail to update to a backdoored composer version.

**Reply by Seldaek** [reply]
I have no idea what you mean, unless someone has access to the machine but then disabling self-update is never going to save you.

**Reply by Seldaek** [reply]
Yes, if someone has remote code exec on your machine you're screwed regardless.. So i don't quite get your point?

**Reply by BSarmady** [reply]
I was just trying to show one of many ways that this can happen bring to light. 

- IdeaJ at the moment always checks for package and self updates automatically when project opens (probably a bug, but exist for few years). the only way to stop IdeaJ from doing it is to move composer.json and ,lock out of project folder and then copy them back when I need them and this will cause auto complete doesn't work for vendor packages.

- Intrusion can happen by package update too and not necessarily just composer self update.

- seeing lots of requests out of a dev pc is enough to freak out the infra team of organization and decide to block all. If we are able to convince them that requests are initiated by user manually and there is nothing automatic, I might be able to get them to agree to leave connection open and just monitor it. This doesn't only apply to your software, I have contacted others too, including VS Code. Some agreed and some ignored and some still thinking, at the end it is up to you if it worth the time to implement such feature.

Anyway, I don't think implementing disable update to be that hard and it is a nice to have feature. I do it in my java, C and C# software easily by adding a file called disable.update in application folder (or where it executes) and when update request comes up, I just look for that file and if exists, I return quietly.

**Reply by uuf6429** [reply]
> I was just trying to show one of many ways that this can happen bring to light.

I'm not a composer developer, but if I were, I'd say we would need concrete reasons to act on this, not what ifs. A reason could be a specific security flaw or some standard people are adopting to have offline-mode applications, but I don't know of either case.

> * IdeaJ at the moment always checks for package and self updates automatically when project opens (probably a bug, but exist for few years). the only way to stop IdeaJ from doing it is to move composer.json and ,lock out of project folder and then copy them back when I need them and this will cause auto complete doesn't work for vendor packages.

My guess is that it does that to show you if there are updates to your packages when you open `composer.json`. If you want to disable that functionality or if you feel that it is a bug, you should remove the Composer settings in the IDE or report the behaviour to JetBrains. Composer cannot decide who can and can't run it.

> * Intrusion can happen by package update too and not necessarily just composer self update.

I mentioned that already...if your/that situation requires that when you run `composer update` nothing happens, then probably you shouldn't even have composer installed.

> * seeing lots of requests out of a dev pc is enough to freak out the infra team of organization and decide to block all. If we are able to convince them that requests are initiated by user manually and there is nothing automatic, I might be able to get them to agree to leave connection open and just monitor it. This doesn't only apply to your software, I have contacted others too, including VS Code. Some agreed and some ignored and some still thinking, at the end it is up to you if it worth the time to implement such feature.

You're basically asking companies / software vendors to cripple their software because your infra team seems insecure and want to micro manage developers. Makes me wonder how they are managing your marketing department - with all those proprietary/closed-source social media apps phoning home to god knows which hosts and for which reason. I.e. in short, micro managing develepers has gone so bad that now you're trying to micro manage software vendors... (in my opinion, this is how it sounds)

> Anyway, I don't think implementing disable update to be that hard and it is a nice to have feature. I do it in my java, C and C# software easily by adding a file called disable.update in application folder (or where it executes) and when update request comes up, I just look for that file and if exists, I return quietly.

I still can't see what you're trying to achieve there. If any user can change/remove that file, then what is the point of it??

Composer is not a daemon or a server...**it needs to be invoked manually**. If I run `composer update` I expect it to update possibly by pigeon mail if it comes to that, period.
Why go through the extra step of saying "won't update because of file or flag" and then the user needs to manually remove that, run composer update again and then add the flag/file back. It's 3 extra steps for absolutely no reason.

As to your feature, to me at first glance, it sounds like a bug, especially when you say "silently return".

Let's say somehow I already had that file somewhere locally. I download composer via cli, try to install some package, everything looks fine (no errors or failures), but the package is not there. I wouldn't even remotely think of looking for a "disable.update" file. (and assuming that I do figure out what's wrong, probably by finding it mentioned in bug reports, I'd say I would lose some respect for composer, because it sounds like a weird/unexpected/pointless behaviour)
What I'm trying to say is that what sounds like a simple feature to you might be a big headache to other users and the developers of composer, so it needs to be thought and implemented the right way.

On the other hand, if there is a standard way to achieve this, then that gives it more credibility / sense. Maybe something like a "--dry-run" or "--offline" flag (although I still wouldn't expect this to be the default behaviour).

**Reply by stof** [reply]
Based on this discussion, it looks like the "automatic" updates @BSarmady is complaining about are the checks done by PHPStorm. But then, this should be reported to PHPStorm, not to Composer.

**Reply by BSarmady** [reply]
@uuf6429 I think this discussion is already over. You are focusing on very minor details to refuse that this is actually an issue. for my team we just set packgist in our dns to local repo which if a package is missing from there it will automatically download it. in my own pc I added to hosts file as localhost so it won't even bother me again. YES, I do work offline, I don't have access to internet all the time specially on weekends when I'm in my barn 40KM away from civilization.

@stof , that is already reported to jetbrain, I couldn't find the case in youtrack, I would link it here. 

Again I quote myself 

> At the end it is up to developer if it worth the time to implement such feature.

**Reply by uuf6429** [reply]
> You are focusing on very minor details to refuse that this is actually an issue.

Yeah, It's this new thing called being pragmatic.

Why should the Composer developers spend time implementing an **ugly hack**, even if it is 2 lines of code, when you can't be bothered to explain your real need? **Especially when said hack will cause trouble and bug reports down the line.**

> for my team we just set packgist in our dns to local repo which if a package is missing from there it will automatically download it. in my own pc I added to hosts file as localhost so it won't even bother me again.

We mentioned this many times already; Composer DOES NOT RUN BY ITSELF. You need to fix whatever is calling composer.
**If this is an IDE problem, then go find the youtrack ticket or report your problem there.**

By the way, I've been using PhpStorm and IDEA Ultimate for over 7 years (in different companies, by myself, on several platforms and on many occasions, offline while on trains). I never had a single problem with the composer integration.

> YES, I do work offline, I don't have access to internet all the time specially on weekends when I'm in my barn 40KM away from civilization.

I'm also working offline, never had a problem with that. The IDE does have an offline mode, did you try that?

> I think this discussion is already over.

Indeed it is.

As a customer, you should know that what you need (offline mode) is not always what you want (a hack in composer).
And as a developer, you should know better how to report problems. Imagine someone in your company comes to you and tells you "hey, from tomorrow on we start coding with pen and paper" - you will probably question that decision and why it's necessary, even if it's super easy to implement (just a pen and paper, right?).

**Reply by stof** [reply]
There is **no** autoupdating in Composer.

**Reply by stof** [reply]
That's exactly why we are asking the submitter to explain in more details what they exactly want to disable (without getting clear answers) as all things we got in response until now are things that don't exist in Composer itself.

**Reply by BSarmady** [reply]
> I'm not a composer developer, but if I were, I'd say we would need

@uuf6429 ,then why are you bothering yourself and take responsibility and talk for someone else's, please allow them to talk themselves.

@stof, 
I hope you are developer of the composer and leading a conversation with no point on behalf of someone else.
phpStorm as soon as project opens, calls the composer update (even if it is disabled in phpStorm settings, it is a long lasting bug already in youtrack)
when this happens, first, composer tries to update itself. I believe this call always happens and there is no way to say I only want to use e.g. composer 4.3 and don't care about 4.4. unless it is just a telemetry call which need its own discussion.
Then after that for as many as packages are in the project, calls are happens to repo even if package version is fixed e.g. `1.4` instead of `^1.4`
I can verify that in the firewall log. 

Also if I'm not mistaken, (have to try it) every time a dev runs composer update on a project, there is an self update call before doing package update.

In a large project with many packages multiply that to the developers opening that project early morning and infra team sees that many requests in firewall log.

**Reply by stof** [reply]
> I hope you are developer of the composer

I am one of the active contributors to composer, even though I'm not officially a maintainer myself (and I know the maintainers)

> phpStorm as soon as project opens, calls the composer update (even if it is disabled in phpStorm settings, it is a long lasting bug already in youtrack)

There is **nothing** composer can do to prevent **phpstorm** from triggering it. If composer were blocking triggers done by PHPStorm automatically, it would still have to provide flags to force running it otherwise PHPStorm users will complain that they cannot use the composer integration anymore, and then PHPStorm could always pass those flags, defeating the purpose. That's why disabling only _automatic_ update checks of PHPStorm can only be achieved by patching PHPStorm and not Composer.
So what you need to do here is to make Jetbrains fix that bug instead of asking Composer developers to try to block this bug (as this cannot be done without breaking many other workflows).

> when this happens, first, composer tries to update itself.

This is totally wrong. Composer **never** attempts to update itself automatically (unless you use a custom build of composer patched to do it or you enabled a composer plugin that runs `self-update` automatically).
The only case where composer is updating itself is when you run the `self-update` command (when running the `diagnose` command, Composer will check whether a new version is available, as that's part of the info displayed in the diagnostic)

> Also if I'm not mistaken, (have to try it) every time a dev runs composer update on a project, there is an self update call before doing package update.

There is no such self update.

> Then after that for as many as packages are in the project, calls are happens to repo even if package version is fixed e.g. `1.4` instead of `^1.4`

When you run `composer update`, composer will indeed load the metadata of packages from packagist.org. But that's the whole point of composer. If it cannot load the metadata, it cannot do its work without those metadata.

Note that when doing _partial_ updates, any package that is kept locked to its current version instead of being part of the update request does not have its metadata loaded from Packagist (unless there is a failure involving such package, in which case composer will load this particular package to be able to give a more detailed error message explaining that the repository has more versions available so that a wider list of updated packages might work).

However, composer never runs such update automatically. Looking at your sentence saying "multiply that to the developers opening that project early morning", it seems once again to me that the *only* cause of the issue is the automatic update done by **PhpStorm**. And as said multiple times before, that's something that must be solved by Jetbrains, not by the Composer team.

**Reply by BSarmady** [reply]
@stof, 

I mentioned that, this is indeed [a phpStorm issue](https://youtrack.jetbrains.com/issue/WI-52582). unfortunately it is a long standing bug from phpStorm. If you look at the issue, there is a checkbox supposedly should stop running composer update, but it doesn't work.

My suggestion wasn't a flag to set to let composer to update, but have a flag to disable the update. 

In this case the flag (that is not set by default) will allow a developer intentionally set it to disable updates.
For users who didn't set the flag everything will be as usual.

I also suggested setting the flag in a file such as composer.settings saved to same place as composer.json and .lock file (similar concept to gradle.settings which is saved to project folder) to make it go offline.

With gradle, if an action requires online resources that are not available in offline mode, a message is shown that "Gradle requires an online resource that is not available in offline mode. Set in gradle.settings, remove the flag to allow completion of task.". Depending on how critical is the task, it might fail remaining tasks too or continue the remaining tasks.

Another minor benefit is that having this flag will make CI/CD pipline script a lot simpler (run two pipelines with same script), one with and one without update which allows running tests automatically in same branch before and after package update and compare results.

**Reply by stof** [reply]
> Another minor benefit is that having this flag will make CI/CD pipline script a lot simpler (run two pipelines with same script), one with and one without update which allows running tests automatically in same branch before and after package update and compare results.

This assumes that your pipeline does not treat a failure of `composer update` as a failure of the script, which looks bad to me. You would rather need to have this check done to skip the call to `composer update` entirely. And so something managed in your pipeline config and not inside the composer call itself.

> If you look at the issue, there is a checkbox supposedly should stop running composer update, but it doesn't work.

If the checkbox does not have any effect, I suggest you to open an issue to report it to PHPStorm (the issue you linked is marked as closed since 2020).

Btw, you should also check whether you don't have a PHPStorm plugin doing additional composer checks instead of being the one controlled by this checkbox.

> My suggestion wasn't a flag to set to let composer to update, but have a flag to disable the update.
> 
> In this case the flag (that is not set by default) will allow a developer intentionally set it to disable updates. For users who didn't set the flag everything will be as usual.

If devs have to remove the flag before being able to intentionally run an update, there is a good chance that they will forget to re-add the flag after that.
Also, note that composer already supports an env variable to disable network on a best effort basis (if the requested content is in the composer cache, the cached value will be used, otherwise things will fail): https://getcomposer.org/doc/03-cli.md#composer-disable-network
